packages/protocol/contracts/L1/verifiers/SGXVerifier.sol

```diff
@@ -6,39 +6,154 @@
+    error SGX_INVALID_INSTANCE();
+    error SGX_INVALID_PROOF_SIZE();
+    error SGX_INSTANCE_REGISTERED();
 
     /// @notice Initializes the contract with the provided address manager.
     /// @param _addressManager The address of the address manager contract.
     function init(address _addressManager) external initializer {
         EssentialContract._init(_addressManager);
     }
 
+    /// @notice Adds trusted SGX instances to the registry.
+    /// @param trustedInstances The address array of trusted SGX instances.
+    function registerInstance(address[] memory trustedInstances)
+        external
+        onlyOwner
+    {
+        for (uint256 i; i < trustedInstances.length; i++) {
+            _replaceInstance(address(0), trustedInstances[i]);
+        }
+    }
+
+    /// @notice Removes trusted SGX instances from the registry.
+    /// @param oldInstance The address of a compromised SGX instance.
+    function invalidateInstance(address oldInstance) external onlyOwner {
+        instances[oldInstance] = 1;
+    }
+
+    /// @notice Adds trusted SGX instances to the registry by another SGX
+    /// instance.
+    /// @param newAddress The new address of the instance.
+    /// @param trustedInstances The address array of trusted SGX instances.
+    /// @param signature The signature proving authenticity.
+    function registerBySgxInstance(
+        address newAddress,
+        address[] memory trustedInstances,
+        bytes memory signature
+    )
+        external
+    {
+        bytes32 signedHash = keccak256(
+            abi.encode("REGISTER_SGX_INSTANCE", newAddress, trustedInstances)
+        );
+        // Would throw in case invalid
+        address signer = signedHash.recover(signature);
+
+        for (uint256 i; i < trustedInstances.length; i++) {
+            _replaceInstance(address(0), trustedInstances[i]);
+        }
+
+        _replaceInstance(signer, newAddress);
+    }
+
     /// @inheritdoc IVerifier
     function verifyProof(
-        // blockId is unused now, but can be used later when supporting
-        // different types of proofs.
         uint64,
-        address,
-        bool,
-        TaikoData.BlockEvidence calldata
+        address prover,
+        bool isContesting,
+        TaikoData.BlockEvidence calldata evidence
     )
         external
+    {
+        // Do not run proof verification to contest an existing proof
+        if (isContesting) return;
+
+        // Size is: 85 bytes
+        // 20 bytes + 65 bytes (signature) = 85
+        if (evidence.proof.length != 85) {
+            revert SGX_INVALID_PROOF_SIZE();
+        }
+
+        address newInstance =
+            address(bytes20(LibBytesUtils.slice(evidence.proof, 0, 20)));
+
+        bytes memory signature = LibBytesUtils.slice(evidence.proof, 20);
+
+        address oldInstance =
+            getSignedHash(evidence, prover, newInstance).recover(signature);
+
+        _replaceInstance(oldInstance, newInstance);
```

Comment:
I believe https://github.com/taikoxyz/taiko-mono/pull/14911/commits/1dd22238278415ec94e3e1d0aa8876806502177a is clearer

Reply:
Can do that, but from gas perspective i still favors the 'id' based solution as Brecht pointed out too. https://github.com/taikoxyz/taiko-mono/pull/14851#discussion_r1350934563

When there is a new instance (and it will be relatively frequent) there will be a 20k gas cost to write them, otherwise if the instance is associated with and ID
then the ID's place in the mapping would be modified:
```
struct InstanceData {
        address activeAddress;
        uint256 setAt;
    }

mapping(uint256 instanceId => address sgxInstance) public sgxRegistry;
```

Very hard to work into the same direction if i constantly get 2 different ones. Please make your call on the design.